### PR TITLE
Mapping engine cleanup

### PR DESCRIPTION
**Pull request by jbogard**
Working version of #1031

### CODE REVIEW

File: src/AutoMapper/Internal/IMappingOptionsMemberMapper.cs [hunk 1]
```diff
@@ -289,7 +289,7 @@ public class NameSplitMember : IChildMemberConfiguration
         public INamingConvention SourceMemberNamingConvention { get; set; }
         public INamingConvention DestinationMemberNamingConvention { get; set; }
 
-        public IEnumerable<MethodInfo> SourceExtensionMethods => (Mapper.Configuration as ConfigurationStore).SourceExtensionMethods;
+        public IEnumerable<MethodInfo> SourceExtensionMethods => (Mapper.Configuration as MapperConfiguration).SourceExtensionMethods;
```
Comment (lbargaoanu) [inline src/AutoMapper/Internal/IMappingOptionsMemberMapper.cs]: I think a cast would work better here. Let it crash if it's the wrong type.

File: src/AutoMapper/IMappingEngine.cs [hunk 2]
```diff
@@ -172,11 +160,20 @@ public interface IMappingEngine
         /// <param name="destination"></param>
         /// <param name="sourceType">Source type to use</param>
         /// <param name="destinationType">Destination type to use</param>
-        [Obsolete("Set the CreateMissingTypeMaps property on Mapper.ConfigurationProvider or your Profile instead")]
         void DynamicMap(object source, object destination, Type sourceType, Type destinationType);
 
+        /// <summary>
+        /// Configuration provider for performaing maps
+        /// </summary>
+        IConfigurationProvider ConfigurationProvider { get; }
```
Comment (lbargaoanu) [inline src/AutoMapper/IMappingEngine.cs]: Why expose this? I think the Mapper should be read-only after CreateMapper.

Comment (TylerCarlson1) [inline src/AutoMapper/IMappingEngine.cs]: ^

Comment (jbogard) [inline src/AutoMapper/IMappingEngine.cs]: It is, the IConfigurationProvider interface is now complete segregated from the mutable side.

**Comment by TylerCarlson1** [comment]
https://github.com/AutoMapper/AutoMapper/blob/develop/src/AutoMapper/Mappers/ExpressionMapper.cs#L211 & https://github.com/AutoMapper/AutoMapper/blob/develop/src/AutoMapper/Mappers/ExpressionMapper.cs#L272 is problem for query tests failing.  Need to pass IMapper or IConfigurationProvider into constructor to fix.

**Comment by lbargaoanu** [comment]
See the comments for [this post](http://blogs.msdn.com/b/pfxteam/archive/2012/02/04/10264111.aspx) about GetOrAdd.

**Comment by jbogard** [comment]
@TylerCarlson1 that did it. The build is passing. One weirdness is the LINQ extension methods will require you to pass in the configuration objects:

```
var config = new MapperConfiguration(cfg => cfg.CreateMap<Source, Dest>());
var builder = config.CreateExpressionBuilder();
_dest = values.ProjectTo<Dest>(builder).ToArray();
```

That builder caches expressions, so I didn't want to combine it with the IMapper/IMappingEngine interface as those might have different lifecycles. IMapper you could actually do one per scoped context, but the expression builder you can cache longer, statically if you want.

I kinda expect people to build their own overloads, and I didn't want to provide a built-in way of doing so I don't think?

I dunno, should I provide a way to expose a setter for an expression builder to not force it to be supplied?

**Comment by TylerCarlson1** [comment]
If you want this in 4.2 or the 4.x series I think it needs to work as is without having to change code as much as possible.  That's what I tried to do with conventions making both the old system work exactly how you expect them to.  Obviously my case sensitivity change was something I didn't make backwards compatible and that had maybe 5-6 issues pop up from that.
That's my limited knowledge on numbering versions.  5.x is when you remove static completely and code using old API intentionally breaks and has to be rewritten.  This probably won't be the case as something always breaks with big changes, but something to strive for.

As for making passing the builder into extensions, I would suggest making `Mapper.Initialize ` set this function for the setter build property.  We are pushing `Mapper.Initialize` to fix build problems, so it should set the expression builder so you don't have to pass it.  I don't know if you should expose it publicly or not.  Maybe the function where you don't pass it in is supported only supported from the Static API, and they have to make their own extension methods if they want to use an instance builder.  The extension where you don't pass the expression builder should also be obsolete and tell the user to use instance

I don't see any good way to satisfy everyone.  I would say static based extensions still work as intended, though Obsolete, and the new extensions you have to pass in the expression builder.

I have to think about it more and look at the code.

Another thing.  I think expression builder should come off of mapper instead of configuration provider.  This is because you only generate one immutable object off of the configuration, where if there was two objects you could possibly make them at two different times.

This is more of a personal preference, I think, and I haven't put as much thought into how this will effect things, as far as refactoring code goes.

**Comment by jbogard** [comment]
btw I just merged this because I wanted to have a chance to give the API a try.

I started down the path of exposing all of those static funcs directly, but I can certainly keep them and mark them as obsolete.

In fact, I'll do that.

**Comment by jbogard** [comment]
And as for the two objects, mapper and expression builder, the reason I did that was because those I thought might have different lifecycles. You can use an IMapper instance with a nested/child container, but with the LINQ stuff you don't really. I imagined IConfiguration/IMapper the same as ISessionFactory/Configuration/ISession with NHibernate. The LINQ stuff I thought could even live off of Configuration, because it only really depends on the configuration anyway. Maybe that's the answer? It would be the difference between `IConfigurationProvider.CreateExpressionBuilder()` and `ProjectTo<Foo>(config)`

But it is a little awkward for sure.

**Comment by TylerCarlson1** [comment]
Oh I thought IConfigurationProvider was mutable, but since that's not the case my point is invalid.

**Comment by jbogard** [comment]
The instance still is mutable underneath the covers. There's just no API exposed to do so. You can only ask it for type maps.

**Comment by TylerCarlson1** [comment]
I don't have a problem with it being under IConfigurationProvider, that's just a matter of personal preference at this point.  I have a problem with the configuration being mutable.

**Pull request merged**